File: app/core/smart_reframe/smoother.py
"""
Camera path smoothing for jitter-free virtual camera motion.

This module provides temporal smoothing of camera positions
to create smooth, professional-looking reframing.
"""

import logging
from typing import Optional

import numpy as np

from app.core.smart_reframe.models import (
    CameraKeyframe,
    CameraMode,
    ShotCameraPlan,
    Shot,
    ShotDetections,
    BoundingBox,
)
from app.core.smart_reframe.config import IntelligentCropConfig
from app.core.smart_reframe.saliency import SaliencyEstimator

logger = logging.getLogger(__name__)


class CameraSmoother:
    """
    Smooth camera paths to eliminate jitter and enforce motion constraints.
    """

    def __init__(self, config: IntelligentCropConfig, fps: float):
        self.config = config
        self.fps = fps
        self.saliency = SaliencyEstimator()

    def compute_camera_plan(
        self,
        shot: Shot,
        detections: ShotDetections,
        frame_width: int,
        frame_height: int,
        video_path: Optional[str] = None,
    ) -> ShotCameraPlan:
        """
        Compute a smooth camera plan for a shot.

        Args:
            shot: The shot to process.
            detections: Detections for this shot.
            frame_width: Source video width.
            frame_height: Source video height.
            video_path: Path to video for saliency fallback.

        Returns:
            ShotCameraPlan with smoothed keyframes.
        """
        # Generate raw focus points from detections
        raw_keyframes = self._compute_raw_focus(
            shot, detections, frame_width, frame_height, video_path
        )

        if not raw_keyframes:
            # Fallback to center if no focus points
            return ShotCameraPlan(
                shot_id=shot.id,
                mode=CameraMode.STATIC,
                keyframes=[
                    CameraKeyframe(
                        time=shot.start_time,
                        cx=frame_width / 2,
                        cy=frame_height / 2,
                        width=frame_width * 0.6,
                        height=frame_height * 0.6,
                    )
                ],
            )

        # Determine camera mode based on motion
        mode = self._classify_camera_mode(raw_keyframes)

        # Apply smoothing
        if mode == CameraMode.STATIC:
            smoothed = self._smooth_static(raw_keyframes)
        else:
            smoothed = self._smooth_tracking(raw_keyframes)

        # Enforce motion constraints
        constrained = self._enforce_constraints(smoothed, frame_width, frame_height)

        return ShotCameraPlan(
            shot_id=shot.id,
            mode=mode,
            keyframes=constrained,
        )

    def _compute_raw_focus(
        self,
        shot: Shot,
        detections: ShotDetections,
        frame_width: int,
        frame_height: int,
        video_path: Optional[str] = None,
    ) -> list[CameraKeyframe]:
        """
        Compute raw focus points from detections.
        """
        # Group detections by time
        time_to_dets: dict[float, list] = {}
        for det in detections.detections:
            if shot.start_time <= det.time <= shot.end_time:
                if det.time not in time_to_dets:
                    time_to_dets[det.time] = []
                time_to_dets[det.time].append(det)

        keyframes = []
        sample_interval = 1.0 / self.config.fps_sample

        # Generate keyframes at regular intervals
        t = shot.start_time
        while t <= shot.end_time:
            # Find closest detection time
            closest_time = None
            min_diff = float("inf")
            for det_time in time_to_dets:
                diff = abs(det_time - t)
                if diff < min_diff and diff < sample_interval:
                    min_diff = diff
                    closest_time = det_time

            if closest_time is not None:
                dets = time_to_dets[closest_time]
                focus = self._compute_focus_from_detections(
                    dets, frame_width, frame_height
                )
                keyframes.append(
                    CameraKeyframe(
                        time=t,
                        cx=focus.cx,
                        cy=focus.cy,
                        width=focus.width,
                        height=focus.height,
                    )
                )
            else:
                # No detection - use fallback
                focus = self.saliency.get_focus_region(
                    np.zeros((frame_height, frame_width, 3), dtype=np.uint8),
                    self.config.fallback_policy,
                )
                keyframes.append(
                    CameraKeyframe(
                        time=t,
                        cx=focus.cx,
                        cy=focus.cy,
                        width=focus.width,
                        height=focus.height,
                    )
                )

            t += sample_interval

        return keyframes

    def _compute_focus_from_detections(
        self,
        detections: list,
        frame_width: int,
        frame_height: int,
    ) -> BoundingBox:
        """
        Compute focus region from a set of detections.
        """
        if not detections:
            return BoundingBox(
                x=frame_width * 0.2,
                y=frame_height * 0.2,
                width=frame_width * 0.6,
                height=frame_height * 0.6,
            )

        if self.config.prefer_primary_subject and len(detections) > 1:
            # Use the largest/most confident detection
            primary = max(detections, key=lambda d: d.bbox.area * d.score)
            focus_box = primary.bbox.pad(
                primary.bbox.width * self.config.subject_padding
            )
        else:
            # Combine all detections
            boxes = [d.bbox for d in detections]
            combined = BoundingBox.union(boxes)
            if combined:
                focus_box = combined.pad(combined.width * self.config.subject_padding)
            else:
                focus_box = BoundingBox(
                    x=frame_width * 0.2,
                    y=frame_height * 0.2,
                    width=frame_width * 0.6,
                    height=frame_height * 0.6,
                )

        return focus_box.clamp(frame_width, frame_height)

    def _classify_camera_mode(self, keyframes: list[CameraKeyframe]) -> CameraMode:
        """
        Classify the camera mode based on motion analysis.
        """
        if len(keyframes) < 2:
            return CameraMode.STATIC

        # Compute motion statistics
        cx_values = [kf.cx for kf in keyframes]
        cy_values = [kf.cy for kf in keyframes]
        width_values = [kf.width for kf in keyframes]

        cx_range = max(cx_values) - min(cx_values)
        cy_range = max(cy_values) - min(cy_values)
        width_range = max(width_values) - min(width_values)

        # Compute standard deviations
        cx_std = np.std(cx_values)
        cy_std = np.std(cy_values)
        width_std = np.std(width_values)

        # Thresholds (relative to average values)
        avg_width = np.mean(width_values)
        motion_threshold = avg_width * 0.1  # 10% of focus width
        zoom_threshold = avg_width * 0.15  # 15% of focus width

        if width_std > zoom_threshold:
            return CameraMode.ZOOM
        elif cx_std > motion_threshold or cy_std > motion_threshold:
            return CameraMode.TRACKING
        else:
            return CameraMode.STATIC

    def _smooth_static(
        self, keyframes: list[CameraKeyframe]
    ) -> list[CameraKeyframe]:
        """
        Smooth keyframes for static camera mode (use average position).
        """
        if not keyframes:
            return []

        # Use median for robustness to outliers
        avg_cx = np.median([kf.cx for kf in keyframes])
        avg_cy = np.median([kf.cy for kf in keyframes])
        avg_width = np.median([kf.width for kf in keyframes])
        avg_height = np.median([kf.height for kf in keyframes])

        return [
            CameraKeyframe(
                time=kf.time,
                cx=avg_cx,
                cy=avg_cy,
                width=avg_width,
                height=avg_height,
            )
            for kf in keyframes
        ]

    def _smooth_tracking(
        self, keyframes: list[CameraKeyframe]
    ) -> list[CameraKeyframe]:
        """
        Smooth keyframes for tracking camera mode (apply low-pass filter).
        """
        if len(keyframes) < 3:
            return keyframes

        # Extract arrays
        times = np.array([kf.time for kf in keyframes])
        cx = np.array([kf.cx for kf in keyframes])
        cy = np.array([kf.cy for kf in keyframes])
        width = np.array([kf.width for kf in keyframes])
        height = np.array([kf.height for kf in keyframes])

        # Compute window size in samples
        sample_rate = len(times) / (times[-1] - times[0]) if times[-1] > times[0] else 1
        window_samples = max(3, int(self.config.smoothing_window * sample_rate))
        if window_samples % 2 == 0:
            window_samples += 1

        # Apply moving average
        cx_smooth = self._moving_average(cx, window_samples)
        cy_smooth = self._moving_average(cy, window_samples)
        width_smooth = self._moving_average(width, window_samples)
        height_smooth = self._moving_average(height, window_samples)

        return [
            CameraKeyframe(
                time=float(times[i]),
                cx=float(cx_smooth[i]),
                cy=float(cy_smooth[i]),
                width=float(width_smooth[i]),
                height=float(height_smooth[i]),
            )
            for i in range(len(times))
        ]

    def _moving_average(self, data: np.ndarray, window: int) -> np.ndarray:
        """Apply moving average with edge handling."""
        if len(data) < window:
            return data

        # Pad edges
        pad = window // 2
        padded = np.pad(data, (pad, pad), mode="edge")

        # Compute cumulative sum for efficient moving average
        cumsum = np.cumsum(padded)
        cumsum = np.insert(cumsum, 0, 0)

        result = (cumsum[window:] - cumsum[:-window]) / window
        return result

    def _enforce_constraints(
        self,
        keyframes: list[CameraKeyframe],
        frame_width: int,
        frame_height: int,
    ) -> list[CameraKeyframe]:
        """
        Enforce motion and boundary constraints on keyframes.
        """
        if len(keyframes) < 2:
            return keyframes

        constrained = [keyframes[0]]

        for i in range(1, len(keyframes)):
            prev = constrained[-1]
            curr = keyframes[i]

            dt = curr.time - prev.time
            if dt <= 0:
                constrained.append(curr)
                continue

            # Compute velocity
            dx = curr.cx - prev.cx
            dy = curr.cy - prev.cy
            speed = np.sqrt(dx**2 + dy**2) / dt

            # Limit speed
            if speed > self.config.max_pan_speed:
                scale = self.config.max_pan_speed / speed
                new_cx = prev.cx + dx * scale
                new_cy = prev.cy + dy * scale
            else:
                new_cx = curr.cx
                new_cy = curr.cy

            # Clamp to frame bounds (with margin for crop window)
            margin_x = curr.width / 2
            margin_y = curr.height / 2
            new_cx = np.clip(new_cx, margin_x, frame_width - margin_x)
            new_cy = np.clip(new_cy, margin_y, frame_height - margin_y)

            constrained.append(
                CameraKeyframe(
                    time=curr.time,
                    cx=float(new_cx),
                    cy=float(new_cy),
                    width=curr.width,
                    height=curr.height,
                )
            )

        return constrained


def compute_camera_plans(
    shots: list[Shot],
    all_detections: list[ShotDetections],
    frame_width: int,
    frame_height: int,
    fps: float,
    config: Optional[IntelligentCropConfig] = None,
) -> list[ShotCameraPlan]:
    """
    Compute camera plans for all shots.

    Args:
        shots: List of shots.
        all_detections: Detections for each shot.
        frame_width: Source video width.
        frame_height: Source video height.
        fps: Video frame rate.
        config: Configuration options.

    Returns:
        List of ShotCameraPlan objects.
    """
    if config is None:
        config = IntelligentCropConfig()

    smoother = CameraSmoother(config, fps)

    # Create a map of shot_id to detections
    det_map = {sd.shot_id: sd for sd in all_detections}

    plans = []
    for shot in shots:
        detections = det_map.get(shot.id, ShotDetections(shot_id=shot.id))
        plan = smoother.compute_camera_plan(
            shot, detections, frame_width, frame_height
        )
        plans.append(plan)

    return plans

File: backend/crates/vclip-media/src/intelligent/smoother.rs
//! Camera path smoothing for jitter-free virtual camera motion.
//!
//! This module provides temporal smoothing of camera positions
//! to create smooth, professional-looking reframing.

use super::config::{FallbackPolicy, IntelligentCropConfig};
use super::models::{BoundingBox, CameraKeyframe, CameraMode, Detection, FrameDetections};
use tracing::debug;

/// Camera smoother for creating smooth camera paths.
pub struct CameraSmoother {
    config: IntelligentCropConfig,
    _fps: f64, // Reserved for future use in timing calculations
}

impl CameraSmoother {
    /// Create a new camera smoother.
    pub fn new(config: IntelligentCropConfig, fps: f64) -> Self {
        Self { config, _fps: fps }
    }

    /// Compute a smooth camera plan from detections.
    ///
    /// # Arguments
    /// * `detections` - Face detections per frame
    /// * `width` - Frame width
    /// * `height` - Frame height
    /// * `start_time` - Start time in seconds
    /// * `end_time` - End time in seconds
    ///
    /// # Returns
    /// Smoothed camera keyframes
    pub fn compute_camera_plan(
        &self,
        detections: &[FrameDetections],
        width: u32,
        height: u32,
        start_time: f64,
        end_time: f64,
    ) -> Vec<CameraKeyframe> {
        // Generate raw focus points from detections
        let raw_keyframes = self.compute_raw_focus(detections, width, height, start_time, end_time);

        if raw_keyframes.is_empty() {
            // Fallback to center if no focus points
            return vec![CameraKeyframe::centered(start_time, width, height)];
        }

        // Determine camera mode based on motion
        let mode = self.classify_camera_mode(&raw_keyframes);
        debug!("Camera mode: {:?}", mode);

        // Apply smoothing based on mode
        let smoothed = match mode {
            CameraMode::Static => self.smooth_static(&raw_keyframes),
            CameraMode::Tracking | CameraMode::Zoom => self.smooth_tracking(&raw_keyframes),
        };

        // Enforce motion constraints
        self.enforce_constraints(&smoothed, width, height)
    }

    /// Compute raw focus points from detections.
    fn compute_raw_focus(
        &self,
        detections: &[FrameDetections],
        width: u32,
        height: u32,
        start_time: f64,
        end_time: f64,
    ) -> Vec<CameraKeyframe> {
        let sample_interval = 1.0 / self.config.fps_sample;
        let mut keyframes = Vec::new();

        let mut current_time = start_time;
        let mut frame_idx = 0;

        while current_time < end_time && frame_idx < detections.len() {
            let frame_dets = &detections[frame_idx];

            let keyframe = if frame_dets.is_empty() {
                // No detection - use fallback
                self.create_fallback_keyframe(current_time, width, height)
            } else {
                // Compute focus from detections
                let focus = self.compute_focus_from_detections(frame_dets, width, height);
                CameraKeyframe::new(
                    current_time,
                    focus.cx(),
                    focus.cy(),
                    focus.width,
                    focus.height,
                )
            };

            keyframes.push(keyframe);
            current_time += sample_interval;
            frame_idx += 1;
        }

        keyframes
    }

    /// Compute focus region from detections.
    fn compute_focus_from_detections(
        &self,
        detections: &[Detection],
        width: u32,
        height: u32,
    ) -> BoundingBox {
        if detections.is_empty() {
            return self.create_fallback_box(width, height);
        }

        if self.config.prefer_primary_subject && detections.len() > 1 {
            // Use the largest/most confident detection
            let primary = detections
                .iter()
                .max_by(|a, b| {
                    let score_a = a.bbox.area() * a.score;
                    let score_b = b.bbox.area() * b.score;
                    score_a.partial_cmp(&score_b).unwrap_or(std::cmp::Ordering::Equal)
                })
                .unwrap();

            let focus_box = primary.bbox.pad(primary.bbox.width * self.config.subject_padding);
            return focus_box.clamp(width, height);
        }

        // Check if faces are close enough to combine
        let faces_far_apart = self.are_faces_far_apart(detections, width);

        if !faces_far_apart {
            // Combine all detections
            let boxes: Vec<BoundingBox> = detections.iter().map(|d| d.bbox).collect();
            if let Some(combined) = BoundingBox::union(&boxes) {
                let focus_box = combined.pad(combined.width * self.config.subject_padding);
                return focus_box.clamp(width, height);
            }
        }

        // Faces far apart - use primary
        let primary = detections
            .iter()
            .max_by(|a, b| {
                let score_a = a.bbox.area() * a.score;
                let score_b = b.bbox.area() * b.score;
                score_a.partial_cmp(&score_b).unwrap_or(std::cmp::Ordering::Equal)
            })
            .unwrap();

        let focus_box = primary.bbox.pad(primary.bbox.width * self.config.subject_padding);
        focus_box.clamp(width, height)
    }

    /// Check if faces are far apart.
    fn are_faces_far_apart(&self, detections: &[Detection], width: u32) -> bool {
        if detections.len() < 2 {
            return false;
        }

        // Compute max distance between face centers
        let mut max_distance: f64 = 0.0;

        for i in 0..detections.len() {
            for j in (i + 1)..detections.len() {
                let dx = detections[i].bbox.cx() - detections[j].bbox.cx();
                let dy = detections[i].bbox.cy() - detections[j].bbox.cy();
                let distance = (dx * dx + dy * dy).sqrt();
                max_distance = max_distance.max(distance);
            }
        }

        // Normalize by frame width
        let normalized = max_distance / width as f64;
        normalized > self.config.multi_face_separation_threshold
    }

    /// Create fallback keyframe based on policy.
    fn create_fallback_keyframe(&self, time: f64, width: u32, height: u32) -> CameraKeyframe {
        let focus = self.create_fallback_box(width, height);
        CameraKeyframe::new(time, focus.cx(), focus.cy(), focus.width, focus.height)
    }

    /// Create fallback bounding box based on policy.
    /// 
    /// For podcast-style videos, faces are typically in the upper 30-45% of frame.
    /// The fallback box should be positioned to capture this region.
    fn create_fallback_box(&self, width: u32, height: u32) -> BoundingBox {
        let w = width as f64;
        let h = height as f64;

        match self.config.fallback_policy {
            FallbackPolicy::Center => {
                // Center of frame - good for general content
                BoundingBox::new(w * 0.2, h * 0.2, w * 0.6, h * 0.6)
            }
            FallbackPolicy::UpperCenter => {
                // Upper-center for talking head / podcast style
                // Position focus box so face center is around 35-40% of frame height
                // Box starts at 15% and is 50% tall, so center is at 40%
                BoundingBox::new(w * 0.15, h * 0.15, w * 0.7, h * 0.5)
            }
            FallbackPolicy::RuleOfThirds => {
                // Rule of thirds - upper third intersection
                BoundingBox::new(w * 0.2, h * 0.15, w * 0.6, h * 0.45)
            }
        }
    }

    /// Classify camera mode based on motion.
    fn classify_camera_mode(&self, keyframes: &[CameraKeyframe]) -> CameraMode {
        if keyframes.len() < 2 {
            return CameraMode::Static;
        }

        // Compute motion statistics
        let cx_values: Vec<f64> = keyframes.iter().map(|kf| kf.cx).collect();
        let cy_values: Vec<f64> = keyframes.iter().map(|kf| kf.cy).collect();
        let width_values: Vec<f64> = keyframes.iter().map(|kf| kf.width).collect();

        let cx_std = std_deviation(&cx_values);
        let cy_std = std_deviation(&cy_values);
        let width_std = std_deviation(&width_values);

        let avg_width = mean(&width_values);
        let motion_threshold = avg_width * 0.1;
        let zoom_threshold = avg_width * 0.15;

        if width_std > zoom_threshold {
            CameraMode::Zoom
        } else if cx_std > motion_threshold || cy_std > motion_threshold {
            CameraMode::Tracking
        } else {
            CameraMode::Static
        }
    }

    /// Smooth keyframes for static camera mode.
    fn smooth_static(&self, keyframes: &[CameraKeyframe]) -> Vec<CameraKeyframe> {
        if keyframes.is_empty() {
            return Vec::new();
        }

        // Use median for robustness to outliers
        let cx = median(&keyframes.iter().map(|kf| kf.cx).collect::<Vec<_>>());
        let cy = median(&keyframes.iter().map(|kf| kf.cy).collect::<Vec<_>>());
        let width = median(&keyframes.iter().map(|kf| kf.width).collect::<Vec<_>>());
        let height = median(&keyframes.iter().map(|kf| kf.height).collect::<Vec<_>>());

        keyframes
            .iter()
            .map(|kf| CameraKeyframe::new(kf.time, cx, cy, width, height))
            .collect()
    }

    /// Smooth keyframes for tracking camera mode.
    fn smooth_tracking(&self, keyframes: &[CameraKeyframe]) -> Vec<CameraKeyframe> {
        if keyframes.len() < 3 {
            return keyframes.to_vec();
        }

        // Compute window size in samples
        let duration = keyframes.last().unwrap().time - keyframes.first().unwrap().time;
        let sample_rate = if duration > 0.0 {
            keyframes.len() as f64 / duration
        } else {
            1.0
        };

        let mut window_samples = (self.config.smoothing_window * sample_rate) as usize;
        window_samples = window_samples.max(3);
        if window_samples % 2 == 0 {
            window_samples += 1;
        }

        // Extract arrays
        let cx: Vec<f64> = keyframes.iter().map(|kf| kf.cx).collect();
        let cy: Vec<f64> = keyframes.iter().map(|kf| kf.cy).collect();
        let width: Vec<f64> = keyframes.iter().map(|kf| kf.width).collect();
        let height: Vec<f64> = keyframes.iter().map(|kf| kf.height).collect();

        // Apply moving average
        let cx_smooth = moving_average(&cx, window_samples);
        let cy_smooth = moving_average(&cy, window_samples);
        let width_smooth = moving_average(&width, window_samples);
        let height_smooth = moving_average(&height, window_samples);

        keyframes
            .iter()
            .enumerate()
            .map(|(i, kf)| {
                CameraKeyframe::new(
                    kf.time,
                    cx_smooth[i],
                    cy_smooth[i],
                    width_smooth[i],
                    height_smooth[i],
                )
            })
            .collect()
    }

    /// Enforce motion constraints on keyframes.
    fn enforce_constraints(
        &self,
        keyframes: &[CameraKeyframe],
        width: u32,
        height: u32,
    ) -> Vec<CameraKeyframe> {
        if keyframes.len() < 2 {
            return keyframes.to_vec();
        }

        let mut constrained = Vec::with_capacity(keyframes.len());
        constrained.push(keyframes[0]);

        for i in 1..keyframes.len() {
            let prev = &constrained[i - 1];
            let curr = &keyframes[i];

            let dt = curr.time - prev.time;
            if dt <= 0.0 {
                constrained.push(*curr);
                continue;
            }

            // Compute velocity
            let dx = curr.cx - prev.cx;
            let dy = curr.cy - prev.cy;
            let speed = (dx * dx + dy * dy).sqrt() / dt;

            // Limit speed
            let (new_cx, new_cy) = if speed > self.config.max_pan_speed {
                let scale = self.config.max_pan_speed / speed;
                (prev.cx + dx * scale, prev.cy + dy * scale)
            } else {
                (curr.cx, curr.cy)
            };

            // Clamp to frame bounds
            let margin_x = curr.width / 2.0;
            let margin_y = curr.height / 2.0;
            let clamped_cx = new_cx.max(margin_x).min(width as f64 - margin_x);
            let clamped_cy = new_cy.max(margin_y).min(height as f64 - margin_y);

            constrained.push(CameraKeyframe::new(
                curr.time,
                clamped_cx,
                clamped_cy,
                curr.width,
                curr.height,
            ));
        }

        constrained
    }
}

// === Helper Functions ===

/// Calculate mean of values.
fn mean(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    values.iter().sum::<f64>() / values.len() as f64
}

/// Calculate standard deviation.
fn std_deviation(values: &[f64]) -> f64 {
    if values.len() < 2 {
        return 0.0;
    }
    let avg = mean(values);
    let variance = values.iter().map(|v| (v - avg).powi(2)).sum::<f64>() / values.len() as f64;
    variance.sqrt()
}

/// Calculate median of values.
fn median(values: &[f64]) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    let mut sorted = values.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let mid = sorted.len() / 2;
    if sorted.len() % 2 == 0 {
        (sorted[mid - 1] + sorted[mid]) / 2.0
    } else {
        sorted[mid]
    }
}

/// Apply moving average with edge handling.
fn moving_average(data: &[f64], window: usize) -> Vec<f64> {
    if data.len() < window {
        return data.to_vec();
    }

    let pad = window / 2;
    let mut result = Vec::with_capacity(data.len());

    for i in 0..data.len() {
        let start = if i >= pad { i - pad } else { 0 };
        let end = (i + pad + 1).min(data.len());
        let slice = &data[start..end];
        result.push(slice.iter().sum::<f64>() / slice.len() as f64);
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_moving_average() {
        let data = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let smoothed = moving_average(&data, 3);

        assert_eq!(smoothed.len(), 5);
        assert!((smoothed[1] - 2.0).abs() < 0.01); // (1+2+3)/3 = 2
        assert!((smoothed[2] - 3.0).abs() < 0.01); // (2+3+4)/3 = 3
    }

    #[test]
    fn test_std_deviation() {
        let data = vec![2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0];
        let std = std_deviation(&data);
        assert!((std - 2.0).abs() < 0.1);
    }

    #[test]
    fn test_median() {
        let data = vec![1.0, 3.0, 5.0, 7.0, 9.0];
        assert_eq!(median(&data), 5.0);

        let data2 = vec![1.0, 2.0, 3.0, 4.0];
        assert_eq!(median(&data2), 2.5);
    }

    #[test]
    fn test_camera_mode_classification() {
        let config = IntelligentCropConfig::default();
        let smoother = CameraSmoother::new(config, 30.0);

        // Static - all same position
        let static_kfs = vec![
            CameraKeyframe::new(0.0, 500.0, 500.0, 100.0, 100.0),
            CameraKeyframe::new(1.0, 500.0, 500.0, 100.0, 100.0),
            CameraKeyframe::new(2.0, 500.0, 500.0, 100.0, 100.0),
        ];
        assert_eq!(smoother.classify_camera_mode(&static_kfs), CameraMode::Static);

        // Tracking - position changes
        let tracking_kfs = vec![
            CameraKeyframe::new(0.0, 100.0, 500.0, 100.0, 100.0),
            CameraKeyframe::new(1.0, 300.0, 500.0, 100.0, 100.0),
            CameraKeyframe::new(2.0, 500.0, 500.0, 100.0, 100.0),
        ];
        assert_eq!(smoother.classify_camera_mode(&tracking_kfs), CameraMode::Tracking);
    }
}
